rfd/0028-cluster-config-resources.md

```diff
@@ -0,0 +1,176 @@
+    bool LocalAuth;
+
+    // SessionControlTimeout is the session control lease expiry and defines
+    // the upper limit of how long a node may be out of contact with the auth
+    // server before it begins terminating controlled sessions.
+    int64 SessionControlTimeout;
+}
+```
+
+It is proposed to distribute them in the following fashion:
+
+```proto
+message ClusterMeta {
+    // This resource is not genuine configuration, it would contain only the
+    // common resource metadata.  The metadata name field would be used to
+    // represent the ClusterID.
+}
+
+message SessionRecordingConfig {
+    string SessionRecording;
+    string ProxyChecksHostKeys;
+}
+
+message ClusterNetworkingConfig {
+    int64 ClientIdleTimeout;
+    int64 KeepAliveInterval;
+    int64 KeepAliveCountMax;
+    int64 SessionControlTimeout;
+}
+
+// The already existing AuditConfig is turned into a standalone resource.
+message AuditConfig { ... }
+
+// LocalAuth field is dropped altogether.  This information should be inferred
+// from ClusterAuthPreference.Type.
+
+// DisconnectExpiredCert field is moved into ClusterAuthPreference.
+message ClusterAuthPreference {
+    ...
+    bool DisconnectExpiredCert;
+}
+```
+
+### Data migration
+
+A new step should be added to `migrateLegacyResources` (called during auth
+server initialization) that takes care of migrating the `ClusterConfig` data
+stored in the backend to the new resource structures.
+
+### Restricting to a subset of values of a field
+
+Some users should not be capable of setting a field to any of the otherwise
+accepted values.  It is proposed to satisfy this requirement by introducing new
+RBAC verbs, interpreted as implying additional constraints. Such new verbs
+would be evaluated together with other RBAC verbs in the auth API handlers.
+
+In particular, this is needed for `SessionRecordingConfig` which has two
+fields:
+* `session_recording` with the possible values: `off`, `node`, `node-sync`,
+  `proxy`, `proxy-sync`;
+* a boolean `proxy_checks_host_keys`.
+
+Some users should be subject to the following restrictions:
+1. The value of `session_recording` can be changed only if its original value
+   is one of `off`, `node` or `node-sync`.
+1. The value of `session_recording` can be changed to only one of `off`, `node`
+   or `node-sync`.
+1. The value of `proxy_checks_host_keys` cannot be changed.
+
+This would be achieved by a new RBAC verb `update.no-proxy`. A user could then
+be assigned a role featuring `update.on-proxy` instead of the regular, more
+permissive `update`:
+```
+  allow:
+    rules:
+    - resources: [session_recording_config]
+      verbs: [read, update.no-proxy]
+```
+
+### Additional dynamically configurable resources
+
+In addition to the resources derived from `ClusterConfig`, the resources
+`ClusterAuthPreference` and `PAMConfig` should also adapted to facilitate
+the dynamic configuration workflow.
+
+### Working with a whole cluster configuration
```

Comment:
Overall looks good to me. Can you add a couple of practical examples using tctl and resources that I've seen customers requesting:

* Set up session recording node-sync
* Configuring PAM custom variables
* Setting auth preference to OIDC for Auth0
* Setting U2f second factor and CA attestation
* Setting global session U2F as required

Reply:
Added some simple examples, let me know if this level of detail is sufficient.